docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+
+## Input Operators
+
+Input operators have additional requirements
+
+* The **emitTuples** method is called by platform giving it an opportunity to emit some data. There are some guidelines on how to implement this method that are important
+	* This should not be a blocking method and should return in a reasonable time that is less than the streaming window size (which is 500ms by default).
+	* If the operator needs to interact with external systems to obtain data and this can potentially take a long time, then this should be performed asynchronously in a different thread. Refer to the threading section below for the guidelines when using threading.
+	* The method can emit more than one data tuples.
+
+### Idempotence
+
+Many applications write data to external systems using output operators. To ensure that data is present exactly once in the external system even in a failure recovery scenario, the output operators expect the replayed windows during recovery contain the same data as before the failure. This is called idempotency. Since operators within the DAG are merely responding to input data provided to them by the upstream operators and the input operator has no upstream operator, the responsibility of idempotent replay falls on the input operators.
+
+* For idempotent replay of data, the operator needs to store some meta-information for every window that would allow it to identify what data was sent in that window. This is called the idempotent state.
+	* If the external source of the input operator allows replayability, this could be information such as offset of last piece of data in the window, an identifier of the last piece of data itself or number of data tuples sent.
+	* However if the external source does not allow replayability from an operator specified point, then the entire data sent within the window may need to be persisted by the operator
+* The platform provides a utility called *WindowDataManager* to allow operators to save and retrieve idempotent state every window. Operators should use this to implement idempotency.
+
+## Output Operators
+
+Output operators typically connect to external storage systems such as filesystems, databases or key value stores to store data.
+
+* In some situations, the external systems may not be functioning in a reliable fashion. They may be having prolonged outages or performance problems. If the operator is being designed to work in such environments, it needs to be able to to handle these problems gracefully and not block the DAG or fail. In these scenarios the operator should cache the data into a local store such as HDFS and interact with external systems in a separate thread so as to not have problems in the operator lifecycle thread. The Reconciler pattern and operators that implement the pattern are available in the library for reference.
+
+### End-to-End Exactly Once
```

Comment:
Here too, the top-level organization is somewhat asymmetric: the first part about the Reconciler has no subtitle but the second part does. Not sure how to reorg though ....